pkg/syncer/specsyncer.go

```diff
@@ -173,9 +237,66 @@ func (c *Controller) applyToDownstream(ctx context.Context, gvr schema.GroupVers
 		}
 	}
 
+	if advancedSchedulingFeatureEnabled {
+		specDiffPatch := upstreamObj.GetAnnotations()[workloadv1alpha1.InternalClusterStatusAnnotationPrefix+c.pcluster]
+		if specDiffPatch != "" {
+			upstreamSpec, specExists, err := unstructured.NestedFieldCopy(upstreamObj.UnstructuredContent(), "spec")
+			if err != nil {
+				return err
+			}
+			if specExists {
+				// TODO(jmprusi): Surface those errors to the user.
+				patch, err := jsonpatch.DecodePatch([]byte(specDiffPatch))
+				if err != nil {
+					klog.Errorf("Failed to decode spec diff patch: %v", err)
+					return err
+				}
+				upstreamSpecJSON, err := json.Marshal(upstreamSpec)
+				if err != nil {
+					return err
+				}
+				patchedUpstreamSpecJSON, err := patch.Apply(upstreamSpecJSON)
+				if err != nil {
+					return err
+				}
+				var newSpec map[string]interface{}
+				if err := json.Unmarshal(patchedUpstreamSpecJSON, &newSpec); err != nil {
+					return err
+				}
+				if err := unstructured.SetNestedMap(downstreamObj.UnstructuredContent(), newSpec, "spec"); err != nil {
+					return err
+				}
+			}
+		}
+	}
+
 	// TODO: wipe things like finalizers, owner-refs and any other life-cycle fields. The life-cycle
 	//       should exclusively owned by the syncer. Let's not some Kubernetes magic interfere with it.
 
+	// TODO(jmprusi): When using syncer virtual workspace we would check the DeletionTimestamp on the upstream object, instead of the DeletionTimestamp annotation,
+	//                as the virtual workspace will set the the deletionTimestamp() on the location view by a transformation.
+	intendedToBeRemovedFromLocation := upstreamObj.GetAnnotations()[workloadv1alpha1.InternalClusterDeletionTimestampAnnotationPrefix+c.pcluster] != ""
+
+	// TODO(jmprusi): When using syncer virtual workspace this condition would not be necessary anymore, since directly tested on the virtual workspace side.
+	stillOwnedByExternalActorForLocation := upstreamObj.GetAnnotations()[workloadv1alpha1.ClusterFinalizerAnnotationPrefix+c.pcluster] != ""
+
+	if intendedToBeRemovedFromLocation && !stillOwnedByExternalActorForLocation {
+		if err := c.toClient.Resource(gvr).Namespace(downstreamNamespace).Delete(ctx, downstreamObj.GetName(), metav1.DeleteOptions{}); err != nil {
+			if errors.IsNotFound(err) {
+				// That's not an error.
+				// Just think about removing the finalizer from the KCP location-specific resource:
```

Comment:
assume the syncer crashes here. How do we recover?

Reply:
Talking to @davidfestal: looks like this is still open. We have to also ensure finalizer removal when the downstream object is "NotFound".